osu.Desktop/GameStateBroadcaster.cs

```diff
@@ -0,0 +1,263 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Net.WebSockets;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Newtonsoft.Json;
+using osu.Framework.Allocation;
+using osu.Framework.Bindables;
+using osu.Framework.Graphics;
+using osu.Framework.Threading;
+using osu.Game.Beatmaps;
+using osu.Game.Configuration;
+using osu.Game.Online.API;
+using osu.Game.Rulesets;
+using osu.Game.Rulesets.Mods;
+using osu.Game.Users;
+
+namespace osu.Desktop
+{
+    internal class GameStateBroadcaster : Component
+    {
+        private readonly List<WebSocketClient> clients = new List<WebSocketClient>();
+
+        private readonly HttpListener listener = new HttpListener();
+
+        private CancellationTokenSource cancellationToken;
+
+        private readonly Bindable<bool> enabled = new Bindable<bool>();
+
+        private readonly GameState state = new GameState();
+
+        private ScheduledDelegate broadcastSchedule;
+
+        private IBindable<User> user;
+
+        public GameStateBroadcaster()
+        {
+            listener.Prefixes.Add("http://localhost:7270/");
+        }
+
+        [BackgroundDependencyLoader]
+        private void load(OsuConfigManager config, IAPIProvider provider, Bindable<RulesetInfo> ruleset, Bindable<WorkingBeatmap> beatmap, Bindable<IReadOnlyList<Mod>> mods)
+        {
+            state.Ruleset.BindTo(ruleset);
+            state.Beatmap.BindTo(beatmap);
+            state.Mods.BindTo(mods);
+
+            user = provider.LocalUser.GetBoundCopy();
+            user.BindValueChanged(u =>
+            {
+                state.Activity.UnbindBindings();
+                state.Activity.BindTo(u.NewValue.Activity);
+            }, true);
+
+            state.Mods.ValueChanged += _ => broadcast();
+            state.Ruleset.ValueChanged += _ => broadcast();
+            state.Beatmap.ValueChanged += _ => broadcast();
+            state.Activity.ValueChanged += _ => broadcast();
+
+            config.BindWith(OsuSetting.PublishGameState, enabled);
+
+            enabled.BindValueChanged(state =>
+            {
+                if (state.NewValue)
+                    Task.Factory.StartNew(start, TaskCreationOptions.LongRunning);
+                else
+                    stop();
```

Comment:
there's no protection against start being called while stop is still running.

stops task is hidden inside a private